rclpy/executors/examples_executors/custom_executor.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+
+from examples_executors.listener import Listener
+from examples_executors.talker import Talker
+import rclpy
+from std_msgs.msg import String
+
+
+class Estopper(rclpy.Node):
+    def __init__(self):
+        super().__init__("estopper")
+        self.sub = self.create_subscription(String, 'estop', self.estop_callback)
+
+    def estop_callback(self, msg):
+        print('I heard: [%s]' % msg.data)
+
+
+class PriorityExecutor(rclpy.Executor):
+    """
+    Execute high priority callbacks in multiple threads, all others in a signle thread.
+
+    This is an example of a custom exectuor in python. Executors are responsible for managing
+    how callbacks get mapped to threads.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.high_priority_nodes = set()
+        self.low_priority_thread = None
+
+    def add_high_priority_node(self, node):
+        self.high_priority_nodes.add(node)
+        # add_node inherited
+        self.add_node(node)
+
+    def can_run_low_priority(self):
+        return self.low_priority_thread is None or not self.low_priority_thread.is_alive()
+
+    def spin_once(self, timeout=None):
+        """
+        Run all available callbacks, then return.
```

Comment:
Ah, now I see that's the case in rclcpp. I confused it with `spin_some`. `spin_once` returns after executing one thing here 4013299